libc/sem.cc

```diff
@@ -76,3 +104,72 @@ int sem_trywait(sem_t* s)
         return libc_error(EAGAIN);
     return 0;
 }
+
+static std::unordered_map<std::string, indirect_semaphore> named_semaphores;
+static mutex named_semaphores_mutex;
+
+OSV_LIBC_API
+sem_t *sem_open(const char *name, int oflag, mode_t mode, unsigned int value)
+{
+    SCOPE_LOCK(named_semaphores_mutex);
+    auto iter = named_semaphores.find(std::string(name));
+    
+    if (iter != named_semaphores.end()) {
+        //opening already named semaphore
+        if (oflag & O_EXCL) {
+            errno = EEXIST;
+            return SEM_FAILED;
+        }
+
+        iter->second.get()->add_reference();
+        return reinterpret_cast<sem_t*>(&(iter->second));
+    }
+    else if (oflag & O_CREAT) {
+        //creating new semaphore
+        if (value > SEM_VALUE_MAX) {
+            errno = EINVAL;
+            return SEM_FAILED;
+        }
+        /*
+         * Ignoring mode bits bits for now. Do we actually want to worry about 
+         * permissions in a unikernel? Would also need to check permissons when 
+         * using open and unlink if so.
+        */
+        named_semaphores.emplace(std::string(name), std::make_unique<posix_semaphore>(value, 1, true)); 
+        return reinterpret_cast<sem_t *>(&named_semaphores[std::string(name)]);
+    }
+    
+    errno = ENOENT;
+    return SEM_FAILED;
+}
+
+OSV_LIBC_API
+int sem_unlink(const char *name)
+{
+    SCOPE_LOCK(named_semaphores_mutex);
+    auto iter = named_semaphores.find(std::string(name));
+    if (iter != named_semaphores.end()) {
+        iter->second.get()->unlink();
+        named_semaphores.erase(std::string(name));
+        if (iter->second.get()->not_referenced()) {
```

Comment:
It's not a good idea (and might even not work?) to use the iterator after having removed the item from the map. Please move this not_referenced check before the erase().